libbeat/publisher/queue/spool/outbroker.go

```diff
@@ -0,0 +1,489 @@
+	"errors"
+	"sync"
+	"time"
+
+	"github.com/elastic/beats/libbeat/publisher"
+	"github.com/elastic/xbeats/txfile/pq"
+)
+
+type outBroker struct {
+	ctx   *spoolCtx
+	state func(*outBroker) bool
+
+	// internal API
+	sigFlushed chan uint
+	get        chan getRequest
+
+	// ack signaling
+	pendingACKs   chanList      // list of pending batches to be forwarded to the ackLoop
+	scheduledACKs chan chanList // shared channel for forwading batches to ackLoop
+	schedACKs     chan chanList // active ack forwarding channel, as used by broker (nil if pendingACKs is empty)
+
+	// queue state
+	queue     *pq.Queue
+	available uint // number of available events. getRequests are only accepted if available > 0
+	events    []publisher.Event
+	required  int
+	total     int
+	active    getRequest
+
+	// internal
+	timer *timer
+	dec   *decoder
+}
+
+type chanList struct {
+	head *ackChan
+	tail *ackChan
+}
+
+type ackChan struct {
+	next  *ackChan
+	ch    chan batchAckMsg
+	total int // total number of events to ACK with this batch
+}
+
+const (
+	// maximum number of events if getRequest size is <0
+	maxEvents = 2048
+)
+
+var ackChanPool = sync.Pool{
+	New: func() interface{} {
+		return &ackChan{
+			ch: make(chan batchAckMsg, 1),
+		}
+	},
+}
+
+var errRetry = errors.New("retry")
+
+func newOutBroker(ctx *spoolCtx, qu *pq.Queue, flushTimeout time.Duration) (*outBroker, error) {
+	b := &outBroker{
+		ctx:   ctx,
+		state: nil,
+
+		// API
+		sigFlushed: make(chan uint, 3),
+		get:        make(chan getRequest),
+
+		// ack signaling
+		pendingACKs:   chanList{},
+		scheduledACKs: make(chan chanList),
+		schedACKs:     nil,
+
+		// queue state
+		queue:     qu,
+		available: qu.Events(),
+		events:    nil,
+		required:  0,
+		total:     0,
+		active:    getRequest{},
+
+		// internal
+		timer: newTimer(flushTimeout),
+		dec:   newDecoder(),
+	}
+
+	b.initState()
+	ctx.Go(b.eventLoop)
+	ctx.Go(b.ackLoop)
+	return b, nil
+}
+
+func (b *outBroker) Consumer() *consumer {
+	return newConsumer(b.ctx, b.get)
+}
+
+// onFlush is run whenever the queue flushes it's write buffer. The callback is
+// run in the same go-routine as the Flush was executed from.
+func (b *outBroker) onFlush(n uint) {
+	if n > 0 {
+		b.sigFlushed <- n
+	}
+}
+
+// onACK is run whenever the queue releases ACKed events. The number of acked
+// events and freed pages will is reported.
+func (b *outBroker) onACK(events, pages uint) {
+}
+
+func (b *outBroker) ackLoop() {
+	log := b.ctx.logger
+
+	log.Debug("start output ack loop")
+	defer log.Debug("stop output ack loop")
+
+	var (
+		activeList chanList
+	)
+
+	for {
+		select {
+		case <-b.ctx.Done():
+			return
+
+		case lst := <-b.scheduledACKs:
+			activeList.concat(&lst)
+
+		case <-activeList.channel():
+			ackCh := activeList.pop()
+
+			for {
+				log.Debugf("receive ACK of %v events\n", ackCh.total)
+				err := b.queue.ACK(uint(ackCh.total))
+				if err != nil {
+					log.Debug("ack failed with:", err)
+					time.Sleep(1 * time.Second)
```

Comment:
My understanding of this part, is we want to give some backoff on the acking when en error occur.
Suggestion, use an exponential backoff with a fixed cap using 1s by default seems a high?

Reply:
Question is how often this could happen. Only reason ACKing fails is due to disk/write failures (network shares).

With exponential backoff we might want to introduce more configurable parameters. I'd prefer not to introduce additional config parameters for dealing with this kind of fatal disk/io issues.

Reply:
OK, good for now.